cmd/imagec/docker.go

```diff
@@ -183,45 +185,89 @@ func FetchImageBlob(options ImageCOptions, image *ImageWithMeta) error {
 	// Open the file so that we can use it as a io.Reader for sha256 calculation
 	in, err := os.Open(string(imageFileName))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer in.Close()
 
-	h := sha256.New()
+	done := make(chan bool)
+	errs := make(chan error)
+	defer close(done)
+	defer close(errs)
+
+	r, w := io.Pipe()
+	t := io.TeeReader(in, w)
+
+	// validate the sha256 sum of the compressed data
+	go func() {
+		defer w.Close()
+		h := sha256.New()
+		progress.Update(po, image.String(), "Verifying Checksum")
+		if _, cerr := io.Copy(h, t); cerr != nil {
+			errs <- cerr
+			return
+		}
+		// Calculate the sum
+		sum := fmt.Sprintf("sha256:%x", h.Sum(nil))
+		if sum != layer {
+			errs <- fmt.Errorf("Failed to validate layer checksum. Expected %s got %s", layer, sum)
+			return
+		}
+		done <- true
+	}()
 
-	progress.Update(po, image.String(), "Verifying Checksum")
-	if _, cerr := io.Copy(h, in); cerr != nil {
-		return cerr
-	}
+	// get the sha256 sum of the decompressed data (DiffID)
+	go func() {
+		defer r.Close()
+		h := sha256.New()
+		tar, cerr := archive.DecompressStream(r)
+		defer tar.Close()
+		if cerr != nil {
+			errs <- cerr
+			return
+		}
+		if _, cerr := io.Copy(h, tar); cerr != nil {
+			errs <- cerr
+			return
+		}
+		d := fmt.Sprintf("%x", h.Sum(nil))
+		diffID = &d
+		done <- true
+	}()
 
-	// Calculate the sum
-	sum := fmt.Sprintf("sha256:%x", h.Sum(nil))
-	if sum != layer {
-		return fmt.Errorf("Failed to validate layer checksum. Expected %s got %s", layer, sum)
+	// wait for both goroutines to finish
```

Comment:
This will work, but `sync.WaitGroup` does this without the need for the `for` with `select`.  You do  the following
```
wg := &sync.WaitGroup{}
go func() { wg.Add(1) }()
go func() { wg.Add(1) }()
wg.Wait()
```